lib/shim/specs/recorder.js

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const SegmentSpec = require('./segment')
+
+/* eslint-disable jsdoc/require-property-description */
+/**
+ * @typedef {object} RecorderSpecParams
+ * @augments SegmentSpecParams
+ * @property {SpecAfterFunction} [after]
+ * @property {number|CallbackBindFunction} [callback]
+ * @property {boolean} [callbackRequired]
+ * @property {boolean} [promise]
+ * @property {number|CallbackBindFunction} [rowCallback]
+ * @property {boolean|string} [stream]
+ */
+
+/**
+ * A callback invoked after an instrumented function has completed its work.
+ * The instrumented function must have been invoked synchronously.
+ *
+ * @typedef {Function} SpecAfterFunction
+ * @param {object} shim The shim used to instrument the external library.
+ * @param {Function} fn The function/method from the external library being
+ * instrumented.
+ * @param {string} name The name of the current segment.
+ * @param {Error|null} error If the instrumented function threw an error, this
+ * will be that error.
+ * @param {*} value The result returned by the instrumented function.
+ * @param {TraceSegment} segment The segment used while instrumenting the
+ * function.
+ */
+
+/**
+ * A specialized case of {@link SegmentSpec}. A `RecorderSpec` is typically
+ * used with {@link Shim.record}. It defines the parameters of segment creation
+ * and segment lifetime.
+ */
+class RecorderSpec extends SegmentSpec {
+  /**
+   * @type {SpecAfterFunction}
+   */
+  after
+
+  /**
+   * If a number, then the number indicates the position in the instrumented
+   * function's arguments list that represents the callback function. Otherwise,
+   * it should be the function to used in conjunction with the instrumented
+   * function.
+   *
+   * @example Using a number
+   * const spec = new RecorderSpec({ callback: -1 })
+   * // elsewhere
+   * const cb = Array.from(arguments).at(spec.callback)
+   *
+   * @example Using a function
+   * const spec = new RecorderSpec({ callback: () => {
+   *   console.log('hello')
+   * })
+   * // elsewhere
+   * instrumentedFunction('foo', spec.callback)
+   *
+   * @type {number|CallbackBindFunction}
+   */
+  callback
+
+  /**
+   * When `true`, a recorded method must be called with a callback for a segment
+   * to be created. Does not apply if a custom callback method has been
+   * assigned via {@link callback}.
+   *
+   * @type {boolean}
+   */
+  callbackRequired
+
+  /**
+   * Indicates if the instrumented function is expected to return a promise.
```

Comment:
just worth noting you could have both callback and promise and it'll be smart and do the right thing. some libraries we instrument either have diff signatures to do both or over time they move towards promise based.